waymo_raw_data_to_argoverse.py

```diff
@@ -30,205 +40,398 @@
+            ) = frame_utils.parse_range_image_and_camera_projection(frame)
+            if args.range_image == 1:
+                (
+                    points_ri,
+                    cp_points_ri,
+                ) = frame_utils.convert_range_image_to_point_cloud(
+                    frame, range_images, camera_projections, range_image_top_pose
+                )
+            elif args.range_image == 2:
+                (
+                    points_ri,
+                    cp_points_ri,
+                ) = frame_utils.convert_range_image_to_point_cloud(
+                    frame,
+                    range_images,
+                    camera_projections,
+                    range_image_top_pose,
+                    ri_index=1,
+                )
+            points_all_ri = np.concatenate(points_ri, axis=0)
+            if args.save_cloud:
+                dump_point_cloud(points_all_ri, timestamp_ns, log_id, ARGO_WRITE_DIR)
+            # Saving labels
+            if args.save_labels:
+                dump_object_labels(
+                    frame.laser_labels,
+                    timestamp_ns,
+                    log_id,
+                    ARGO_WRITE_DIR,
+                    track_id_dict,
+                )
+            if args.save_calibration:
+                calib_json = form_calibration_json(frame.context.camera_calibrations)
+                if log_calib_json is None:
+                    log_calib_json = calib_json
+                    calib_json_fpath = (
+                        f"{ARGO_WRITE_DIR}/{log_id}/vehicle_calibration_info.json"
+                    )
+                    check_mkdir(str(Path(calib_json_fpath).parent))
+                    save_json_dict(calib_json_fpath, calib_json)
+                else:
+                    assert calib_json == log_calib_json
+
+            # 5 images per frame
+            for index, tf_cam_image in enumerate(frame.images):
+                # 4x4 row major transform matrix that transforms
+                # 3d points from one frame to another.
+                SE3_flattened = np.array(tf_cam_image.pose.transform)
+                city_SE3_egovehicle = SE3_flattened.reshape(4, 4)
+                # in seconds
+                timestamp_s = tf_cam_image.pose_timestamp
+                timestamp_ns = int(timestamp_s * 1e9)  # to nanoseconds
+                if args.save_poses:
+                    dump_pose(city_SE3_egovehicle, timestamp_ns, log_id, ARGO_WRITE_DIR)
+
+                if args.save_images:
+                    camera_name = CAMERA_NAMES[tf_cam_image.name]
+                    img = tf.image.decode_jpeg(tf_cam_image.image)
+                    new_img = undistort_image(
+                        np.asarray(img),
+                        frame.context.camera_calibrations,
+                        tf_cam_image.name,
+                    )
+                    img_save_fpath = f"{ARGO_WRITE_DIR}/{log_id}/{camera_name}/{camera_name}_{timestamp_ns}.jpg"
```

Comment:
do we need to round any of these nanosecond timestamps to nearest microseconds? (i.e. 
`t_nanos = 1508103378165379072` becomes `t_micros = 1508103378165379000`? I remember there were weird mismatches when I did this, because of different time types/rounding/precision on their end. We should check if labels and lidar and camera times are all provided in nanoseconds

Reply:
Yeah, that's a good point. Should we just round the timestamp from the camera image to microseconds so we don't have mismatches?

Reply:
Hmm maybe it's ok? Let's see if the LiDAR visualization on images works ok with the current format:

Camera pose timestamp and image timestamp are both doubles: https://github.com/waymo-research/waymo-open-dataset/blob/master/waymo_open_dataset/dataset.proto#L244. They must match for the Argoverse tools to work.

The LiDAR timestamp is a int64 https://github.com/waymo-research/waymo-open-dataset/blob/master/waymo_open_dataset/dataset.proto#L298 
but it's fine if that doesn't match the LiDAR, since it's an entirely different sensor.

Reply:
Yeah so the camera pose and image timestamp are always the same. I saw that `cuboids_to_bboxes` from argoverse finds the closest LIDAR timestamp.